### PR TITLE
Add support for Android view capturing in Session Replay

### PR DESCRIPTION
**Pull request by jonathanmos**
### What does this PR do?
Adds support for capturing Android views in Session Replay. Views are supported with the opacity prop or where specifying the opacity through a color with alpha (e.g. FF000000)

Support for border width and color, and for corner radius.

emulator:
Screenshot 2023-11-26 at 21 28 49
replay:
Screenshot 2023-11-28 at 12 07 48

With border and rounded corners.
emulator:
Screenshot 2023-11-29 at 12 50 55

replay:
Screenshot 2023-11-29 at 12 50 09

With parent with 10% opacity:
emulator
Screenshot 2023-11-28 at 13 12 52

replay
Screenshot 2023-11-28 at 13 12 42

### Motivation
Part of the effort to add Session Replay to React Native.

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        )
+
+        val pixelDensity = mappingContext.systemInformation.screenDensity
+        val backgroundDrawable = view.background as? ReactViewBackgroundDrawable
+
+        val backgroundColor = backgroundDrawable?.color ?: 0
+        val viewAlpha = resolveAlphaValue(view.alpha, backgroundColor)
+
+        if (viewAlpha == 0) {
+            return emptyList()
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: I see 2 corner cases that I'd like to check with this condition:
1. If there is another element inside the view that does not have an alpha of `0`, is it going to be rendered on the screen? if so, is it going to be in the replay as well?
2. If a View has an alpha of `0`, I think it still takes space on the screen? does it on the player as well? (I also think so since I think we use coordinates to place elements)

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    ): List<MobileSegment.Wireframe> {
+        val viewGlobalBounds = resolveViewGlobalBounds(
+            view,
+            mappingContext.systemInformation.screenDensity
+        )
+
+        val pixelDensity = mappingContext.systemInformation.screenDensity
+        val backgroundDrawable = view.background as? ReactViewBackgroundDrawable
+
+        val backgroundColor = backgroundDrawable?.color ?: 0
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: Can you attach a screenshot of a view with no background color?
I'm not sure about setting `0` as default value here, I'm afraid it would make the background all black when it should be transparent by default

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        val viewAlpha = resolveAlphaValue(view.alpha, backgroundColor)
+
+        if (viewAlpha == 0) {
+            return emptyList()
+        }
+
+        val color = colorAndAlphaAsStringHexa(
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: nit: `color` is usually used for font color, maybe we can change it to `backgroundColorAndAlpha`?

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+private const val ALPHA_CODE_LENGTH: Int = 2
+private const val HEX_RADIX: Int = 16
+private const val OPAQUE_ALPHA_VALUE: Int = 255
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveAlphaValue(viewAlpha: Float, backgroundColor: Int): Int {
+    // must not be toHexString because we'll lose the sign bit
+    val bgColorAsHexCode = backgroundColor.toString(HEX_RADIX)
+    var pct = viewAlpha
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: I don't know what `pct` is, could you make the variable name more explicit or add a comment if it is a concept in color?

Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: I'll change it to opacityPercentage - a value between 0 and 1

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt (hunk 2 above)
Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: It appears that 0 is the code for transparency as defined in Color.java. In recent rn versions they've made this more explicit in the sourcecode

File: packages/react-native-session-replay/android/src/test/kotlin/com/datadog/reactnative/sessionreplay/ColorUtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    fun `M use opacity code W resolveAlphaValue { color with opacity }`() {
+        // Given
+        val color = 1717960806
+
+        // When
+        val alphaValue = resolveAlphaValue(0.1f, color)
+
+        // Then
+        assertThat(alphaValue)
+            .isEqualTo(102)
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/src/test/kotlin/com/datadog/reactnative/sessionreplay/ColorUtilsTest.kt]: Could you explain how `102` is the expected value? You can translate from decimal to hex in comments for more clarity

**Comment by louiszawadzki** [comment]
Can you add screenshots of the nigthly app to see what it looks like?

Also I don't see anything around the `opacity` prop, if this is handled by another prop, could you mention it in a comment in the code?

File: packages/react-native-session-replay/android/src/test/kotlin/com/datadog/reactnative/sessionreplay/ColorUtilsTest.kt (hunk 5 above)
Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/test/kotlin/com/datadog/reactnative/sessionreplay/ColorUtilsTest.kt]: I'll add a comment, but basically color code 1717960806 translates to hex 66660066, where 66 is the hex alpha code. [66 hex alpha code is 40%](https://gist.github.com/lopspower/03fb1cc0ac9f32ef38f4) and 40% of 255 is 102 alpha.

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt (hunk 1 above)
Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: With the current changes, 1 is handled and if the parent has an opacity of 0 then the children are not rendered. Regarding the second, we specify wireframes by absolute coordinates so I don't think it will be an issue. Let's discuss a test for this

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
+        if (backgroundDrawable == null) {
+            return MobileSegment.ShapeStyle(
+                backgroundColor = viewColor,
+                opacity = viewAlpha,
+            ) to null
+        } else {
+            val cornerRadius = backgroundDrawable.fullBorderRadius
+            val borderColor = resolveBorderColor(backgroundDrawable, opacity)
+            val borderWidth = backgroundDrawable.fullBorderWidth
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: On the screenshots you shared I'm under the impression that the border appears bigger in the replay than on emulators. 
Could you check that this works properly?
Maybe we're missing `.convertToDensityNormalized(pixelDensity)`?

Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: yes, good catch - I've added conversion to dp

File: packages/core/android/build.gradle [hunk 7]
```diff
@@ -169,10 +169,10 @@ dependencies {
   implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
   compileOnly "com.squareup.okhttp3:okhttp:3.12.13"
   
-  implementation "com.datadoghq:dd-sdk-android-rum:2.2.0"
-  implementation "com.datadoghq:dd-sdk-android-logs:2.2.0"
-  implementation "com.datadoghq:dd-sdk-android-trace:2.2.0"
-  implementation "com.datadoghq:dd-sdk-android-webview:2.2.0"
```
Comment (louiszawadzki) [inline packages/core/android/build.gradle:175]: Can you make this change part of another commit? that will make it easier to revert if it needs to be reverted.

Comment (jonathanmos) [reply packages/core/android/build.gradle:175]: done 👍

**Review by louiszawadzki — DISMISSED** [review]
Works for me, I'd like to have an extra approval from a member of the Android SDK team as well.

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    if (bgColorAsHexCode?.length == HEX_COLOR_INCLUDING_ALPHA_LENGTH) {
+        val opacityCode = bgColorAsHexCode.take(ALPHA_CODE_LENGTH)
+        opacityPercentage = alphaCodeToPct(opacityCode)
+    }
+
+    return (opacityPercentage * OPAQUE_ALPHA_VALUE).toInt()
+}
+
+@VisibleForTesting
+internal fun alphaCodeToPct(hexAlpha: String): Float {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: I think we can avoid using the short abbreviation

```suggestion
internal fun alphaCodeToPercentage(hexAlpha: String): Float {
```

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+import androidx.annotation.VisibleForTesting
+import com.facebook.react.views.view.ReactViewGroup
+
+private const val ALPHA_CODE_LENGTH: Int = 2
+private const val HEX_RADIX: Int = 16
+private const val OPAQUE_ALPHA_VALUE: Int = 255
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveAlpha(opacity: Float, backgroundColor: Int?): Int {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: I'm a bit confused by this method. It means, that if `backgroundColor` has alpha bits set, we completely ignore `opacity`, it is kind of non-obvious behavior. Why this method takes both parameters and not take only `backgroundColor`, extracts alpha and then compares with `opacity` somewhere else?

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    private fun resolveRNShapeStyleAndBorder(
+        backgroundDrawable: ReactViewBackgroundDrawable?,
+        viewAlpha: Int,
+        viewColor: String,
+        opacity: Float,
+        pixelDensity: Float
+    ):
+            Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]:
```suggestion
    ): Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
```

**Comment by jonathanmos** [comment]
I refactored three files -> ColorUtils, ColorUtilsTest and ReactViewGroupMapper. The changes were quite extensive so could you please review those three again? Basically, I separated opacity and color alpha entirely as concepts and implemented a simpler method to calculate the rgba from the color int. I also removed the recursion for the opacity, because this logic is not used on the native side and the solution will either have to be on the player side, or on the native side perhaps will require an rfc

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+package com.datadog.reactnative.sessionreplay
+
+import androidx.annotation.VisibleForTesting
+
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveBackgroundColorAsHex(backgroundColor: Int): String {
+    val colorHexString = Integer.toHexString(backgroundColor)
+
+    // reorder the hex string from argb to rgba
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: I think comment is not needed, method is self-explanatory

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    return "#${reorderARGBtoRGBA(colorHexString)}"
+}
+
+@VisibleForTesting
+internal fun reorderARGBtoRGBA(hexString: String): String {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: maybe just `argbToRgba`? or `convertArgbToRgba`. I guess these names are more common, at least `convert` is the lingvo used in `android.graphics.Color`.

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveBackgroundColorAsHex(backgroundColor: Int): String {
+    val colorHexString = Integer.toHexString(backgroundColor)
+
+    // reorder the hex string from argb to rgba
+    return "#${reorderARGBtoRGBA(colorHexString)}"
+}
+
+@VisibleForTesting
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: it seems we can make this method private and just test `resolveBackgroundColorAsHex`?

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.reactnative.sessionreplay
+
+import androidx.annotation.VisibleForTesting
+
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveBackgroundColorAsHex(backgroundColor: Int): String {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: it is not clear which format will we return. Normally everyone can expect ARGB, but this method returns RGBA. Maybe it makes sense to rename this method to `formatAsRgba` or something similar? Also obviously no need to mention `background` in the name, this method works with any color.

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                viewGlobalBounds.y,
+                viewGlobalBounds.width,
+                viewGlobalBounds.height,
+                shapeStyle = shapeStyle,
+                border = border
+            )
+        )
+    }
+
+    @SuppressLint("VisibleForTests")
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:65]: why this annotation is needed? what is the warning?

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    private fun resolveRNShapeStyleAndBorder(
+        view: ReactViewGroup,
+        backgroundDrawable: ReactViewBackgroundDrawable,
+        opacity: Float,
+        pixelDensity: Float
+    ): Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt]: it seems that return type should be `Pair<MobileSegment.ShapeStyle, MobileSegment.ShapeBorder>`, because the only return path in this method return non-null values for the pair.

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt (hunk 15 above)
Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:65]: The error is "This method should only be accessed from tests or within private scope", returning from view.backgroundColor, because this method is annotated VisibleForTesting

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt (hunk 14 above)
Comment (jonathanmos) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt]: 👍 I'll rename it to formatAsRgba

**Review by 0xnm — APPROVED** [review]

**Review by louiszawadzki — APPROVED** [review]
